File: scripts/utility/json-formatter/javascript/json-formatter.js
#!/usr/bin/env node
/*
 * JSON Formatter - pretty-print, minify and validate JSON.
 *
 * Reads JSON from a file argument or stdin and writes the result to stdout.
 * Exits non-zero on invalid JSON so it composes well in shell pipelines.
 */
'use strict';

const fs = require('fs');

function parseArgs(argv) {
  const opts = { file: null, minify: false, validate: false, sortKeys: false, indent: 2 };
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    switch (arg) {
      case '--minify': opts.minify = true; break;
      case '--validate': opts.validate = true; break;
      case '--sort-keys': opts.sortKeys = true; break;
      case '--indent': opts.indent = parseInt(argv[++i], 10) || 2; break;
      default:
        if (!arg.startsWith('--')) opts.file = arg;
    }
  }
  return opts;
}

function sortDeep(value) {
  if (Array.isArray(value)) return value.map(sortDeep);
  if (value && typeof value === 'object') {
    return Object.keys(value).sort().reduce((acc, key) => {
      acc[key] = sortDeep(value[key]);
      return acc;
    }, {});
  }
  return value;
}

function readInput(file) {
  if (file) return fs.readFileSync(file, 'utf8');
  return fs.readFileSync(0, 'utf8'); // file descriptor 0 = stdin
}

function main() {
  const opts = parseArgs(process.argv.slice(2));

  let raw;
  try {
    raw = readInput(opts.file);
  } catch (err) {
    process.stderr.write(`Error reading input: ${err.message}\n`);
    process.exit(2);
  }

  let data;
  try {
    data = JSON.parse(raw);
  } catch (err) {
    process.stderr.write(`Invalid JSON: ${err.message}\n`);
    process.exit(1);
  }

  if (opts.validate) {
    process.stderr.write('Valid JSON.\n');
    process.exit(0);
  }

  if (opts.sortKeys) data = sortDeep(data);

  const output = opts.minify
    ? JSON.stringify(data)
    : JSON.stringify(data, null, opts.indent);

  process.stdout.write(output + '\n');
}

main();

File: scripts/utility/json-formatter/python/json-formatter.py
#!/usr/bin/env python3
"""JSON Formatter - pretty-print, minify and validate JSON.

Reads JSON from a file or stdin and writes the result to stdout (or back to the
file with --in-place). Exits non-zero on invalid JSON so it composes well in
shell pipelines and CI checks.
"""

from __future__ import annotations

import argparse
import json
import sys
from typing import Optional


def parse_args(argv: Optional[list[str]] = None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    parser.add_argument(
        "file", nargs="?",
        help="JSON file to process (reads from stdin if omitted).",
    )
    parser.add_argument("--minify", action="store_true",
                        help="Output compact single-line JSON.")
    parser.add_argument("--validate", action="store_true",
                        help="Only check validity; print nothing on success.")
    parser.add_argument("--sort-keys", action="store_true",
                        help="Sort object keys alphabetically.")
    parser.add_argument("--indent", type=int, default=2,
                        help="Indent width for pretty output (default: 2).")
    parser.add_argument("--in-place", action="store_true",
                        help="Rewrite the input file with the result.")
    return parser.parse_args(argv)


def main(argv: Optional[list[str]] = None) -> int:
    args = parse_args(argv)

    if args.in_place and not args.file:
        print("Error: --in-place requires a file argument.", file=sys.stderr)
        return 2

    raw = (open(args.file, encoding="utf-8").read() if args.file
           else sys.stdin.read())

    try:
        data = json.loads(raw)
    except json.JSONDecodeError as exc:
        print(f"Invalid JSON: {exc}", file=sys.stderr)
        return 1

    if args.validate:
        print("Valid JSON.", file=sys.stderr)
        return 0

    if args.minify:
        output = json.dumps(data, separators=(",", ":"),
                            sort_keys=args.sort_keys, ensure_ascii=False)
    else:
        output = json.dumps(data, indent=args.indent,
                            sort_keys=args.sort_keys, ensure_ascii=False)

    if args.in_place:
        with open(args.file, "w", encoding="utf-8") as handle:
            handle.write(output + "\n")
        print(f"Formatted {args.file}", file=sys.stderr)
    else:
        print(output)
    return 0


if __name__ == "__main__":
    sys.exit(main())
